### ISSUE TITLE
named swizzles 

### ISSUE THREAD

**Issue by greggman** [gpuweb/gpuweb#4333]
This was probably already discussed and maybe it's too hard but sometimes I really wish I could name the swizzles. For example this code

```wgsl
      fn hsl2rgb(hsl: vec3f) -> vec3f {
        let c = vec3f(fract(hsl.x), clamp(hsl.yz, vec2f(0), vec2f(1)));
        let rgb = clamp(abs((c.x * 6.0 + vec3f(0.0, 4.0, 2.0)) % 6.0 - 3.0) - 1.0, vec3f(0), vec3f(1));
        return c.z + c.y * (rgb - 0.5) * (1.0 - abs(2.0 * c.z - 1.0));
      }
```

I'd really like to able to make a type `HSL` and then write this as

```wgsl
      fn hsl2rgb(hsl: HSL) -> vec3f {
        let c = hsl(fract(hsl.h), clamp(hsl.sl, vec2f(0), vec2f(1)));
        let rgb = clamp(abs((c.h * 6.0 + vec3f(0.0, 4.0, 2.0)) % 6.0 - 3.0) - 1.0, vec3f(0), vec3f(1));
        return c.l + c.s * (rgb - 0.5) * (1.0 - abs(2.0 * c.l - 1.0));
      }
```

As remembering that `c.x` is hue, `c.y` is saturation, and `c.z` is lightness is a PITA and arguably unreadable. I'm not sure it's all that much more readable with `.h`, `.s` and `.l`. 

I might even like longer names than single letters

I can make HSL like this 

```wgsl
struct HSL {
   h: f32,
   s: f32,
   l: f32,
}
```

but what I want is something that stays a `vec3f` but lets me name the accessors

Maybe there's no advantage to the terseness? I could write this

```wgsl
      fn hsl2rgb(hsl: HSL) -> vec3f {
        let hue = fract(hsl.x);
        let saturation = clamp(hsl.y, 0.0, 1.0);
        let lightness = clamp(hsl.x, 0.0, 1.0);
        let rgb = clamp(abs((hue * 6.0 + vec3f(0.0, 4.0, 2.0)) % 6.0 - 3.0) - 1.0, vec3f(0), vec3f(1));
        return lightness + saturation * (rgb - 0.5) * (1.0 - abs(2.0 * lightness - 1.0));
      }
```

Is that always the case or are there plenty of cases where keeping things in a vec3 or vec4 has performance advantages but then I lose any semantic meaning.

I think none of the other shader languages support this, maybe MSL since it has C++ like templates?

I find it a little disappointing so much shader code is so hard to read and just kind of wondering out loud if WGSL could somehow encourage more readable shaders.

**Comment by magcius** [comment]
> Is that always the case or are there plenty of cases where keeping things in a vec3 or vec4 has performance advantages but then I lose any semantic meaning.

If vec3/vec4 had any performance benefit, it would come from the fact that you're doing multiple similar operations across the vector, but that's not at all what's going on here. You're doing independent operations across three portions. So, even on a vector ISA, you would have a performance *loss*.

Thankfully, most GPUs these days are scalar ISAs, and compilers effectively treat vec3's and vec4's as bundles of floats, the same as a struct full of floats. I see no reason to keep using vec3/vec4 here.

**Comment by greggman** [comment]
> but that's not at all what's going on here. You're doing independent operations across three portions. So, even on a vector ISA, you would have a performance _loss_.

So are you saying this is a pointless request because there is no such thing a shader where x,y,z,w represent something other than x,y,z,w? Because that sounds like what you're saying.

**Comment by greggman** [comment]
I feel like you're taking this specific example and ignoring the idea because of a bad example. We have xyzw and rgba. Some languages also have stpq.   Why only those? Why not let the user choose names (or letters) that fit their problem?

**Comment by magcius** [comment]
Yes, I'm responding to your comment asking if you needed to keep things in vec3's. If you used a custom structure as in your last example, it would not affect performance adversely, and would also allow you to use the names you wish to use.

> Why not let the user choose names (or letters) that fit their problem?

We already have structures, and functions, both of which allow this, and are the tool I would recommend in this case.

**Comment by greggman** [comment]
By that argument then we should get rid of swizzles. Users can use structures if they want x,y,z,w or r,g,b,a  You don't need swizzles.  If you do need them, then you arguably need them be decided by the user.

**Comment by magcius** [comment]
I'd say that swizzles are helpful for common simple coordinate transforms. There are often symmetric transforms that can be shared between X and Y, but not Z, for instance, and it's helpful to "pick" xy out but not z. While at one time there was a performance argument to be made in favor of swizzles (and the reason they were introduced), those days are long gone, and so now it remains just for the convenience and terseness factor.

But, as you've noticed, swizzles can lead to confusing code as sometimes the input conventions are not documented. There have been attempts at extending swizzle syntax to bigger types, like the [matrix swizzle found in HLSL](https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-per-component-math#the-matrix-type), but it's pretty awkward, and doesn't show up much in practice.

I'd say the modern style of HLSL is to eschew lots of vector swizzles in favor of structures, and even embrace things like methods:

```hlsl
struct HSL {
   float H;
   float S;
   float L;

    void Normalize() {
        H = fract(H);
        S = saturate(S);
        L = saturate(L);
    }

    float3 GetRGB() {
        Normalize();
        // Unwrap the hue angle around the circle (H) back to the relevant point on the edge of the RGB cube.
        float3 corner = saturate(abs((H * 6.0 + vec3f(0.0, 4.0, 2.0)) % 6.0 - 3.0) - 1.0);
        // Saturation (S) lerps us away black/white line, lightness (L) determines the point on the black/white line.
        return L + S * (corner - 0.5) * (1.0 - abs(2.0 * L - 1.0));
   }
};
```

WGSL doesn't have methods, but I think it shows a different direction that's possible.
